src/Microsoft.AspNetCore.OData/Routing/ODataPathRouteConstraint.cs

```diff
@@ -48,13 +48,36 @@ public virtual bool Match(HttpContext httpContext, IRouter route, string routeKe
                 object oDataPathValue;
                 if (values.TryGetValue(ODataRouteConstants.ODataPath, out oDataPathValue))
                 {
+                    string odataPath = oDataPathValue as string;
+                    Func<IServiceProvider> requestContainerFactory = () =>
+                    {
+                        // Delegate is reused. We need to ensure only one request container is created
+                        IServiceProvider requestContainer = request.ODataFeature().RequestContainer;
+
+                        if (requestContainer != null)
+                        {
+                            return requestContainer;
+                        }
+
+                        return request.CreateRequestContainer(RouteName);
+                    };
+
+                    // Check whether the request is a POST targeted at a resource path ending in /$query
+                    if (request.IsQueryRequest(odataPath))
+                    {
+                        request.TransformQueryRequest(requestContainerFactory);
+
+                        odataPath = odataPath.Substring(0, odataPath.LastIndexOf('/' + ODataRouteConstants.QuerySegment, StringComparison.OrdinalIgnoreCase));
+                        values[ODataRouteConstants.ODataPath] = odataPath;
```

Comment:
What does this do? Or why is it necessary?

Reply:
@odero When `/$query` is appended to the resource path, the `oDataPath` key in the route values dictionary will be mapped to the value `/Users/$query`. Say after parsing the request body we end up with query options string like this: `$filter=Role eq 'Administrator'&$orderby=Id`.

The reconstructed OData URL will look something like this: `/Users?$filter=Role eq 'Administrator'&$orderby=Id`

Note that we drop the `/$query` segment from the URL completely at this point and this is what necessitates updating the route values as well, replacing `/Users/$query` with `/Users`

Reply:
I see so could you just set it as
```c#
values[ODataRouteConstants.ODataPath] = request.Path.Value;
// or set oDataPath as
oDataPath = request.Path.Value;
```

Reply:
@odero The problem here is the `oDataPath` value from the route dictionary is not an exact match of `request.Path` value. For instance, based on the example above, `oDataPath` would initially have the value `Users/$query` while `request.Path` would have the value `/Users/$query` (starts with a /).

Whichever route you take, some string manipulation will be involved. I just found it easier/safer to drop `/$query` segment

Reply:
Here's a screenshot

![image](https://user-images.githubusercontent.com/28291332/84031113-cd724880-a99d-11ea-8819-21f04074b79f.png)

Reply:
Notice that the Request Path includes even the route prefix